VPN-5251 - Implement authentication and subscription interaction telemetry

brizental opened the pull request.

mcleinman commented:
Was there a tech spec for this?

brizental commented:
> Was there a tech spec for this?

On the ticket. I just copy pasted the ticket's description in here too.

brizental commented:
About data review. I am not going to request data review for these metrics yet. The names and categories and extra keys are likely to change in the near future. Me, Santiago and Yeonjoo are likely to reach a decision on it by the end of the week.

mcleinman commented:
> > Was there a tech spec for this?
> 
> On the ticket. I just copy pasted the ticket's description in here too.

Is that mira board the tech spec? When was the review on that?

brizental commented:
> Is that mira board the tech spec? When was the review on that?

Yes, the Miro board is the tech spec, alongside https://docs.google.com/document/d/19ujr0v2rGAOKtVjIDHNj6pcK9OoESxxYKFysig5FGvA/edit?usp=sharing which was presented in one of our engineering meetings by Yeonjoo. About the review, this is more of a Santiago and Yeonjoo question. I did not do the spec, I am just implementing it.

lesleyjanenorton commented on src/ui/authenticationInApp/ViewAuthenticationSignUp.qml:
```diff
@@ -29,6 +36,9 @@ MZInAppAuthenticationBase {
     _inputs: MZInAppAuthenticationInputs {
         objectName: "authSignUp"
 
+        _telemetryScreenId: authSignUp._telemetryScreenId
```
I'm surprised this is necessary. Is `_telemetryScreenId` not inherited from `MZAuthenticationBase`?

brizental replied:
Eh, this could just be my QML ignorance showing really.

brizental replied:
I tried removing it but it is not inherited it seems like.

mcleinman reviewed (changes requested):
A PR of this length deserves a PR description that is more than copy/pasting from the ticket. It's kind to the reviewers to give a bit of background on the code, things to watch out for, some background on decisions that you made, etc.

Gela commented on src/inspector/inspectorhandler.cpp:
```diff
@@ -380,8 +380,6 @@ static QList<InspectorCommand> s_commands{
                        QPointF pointF = item->mapToScene(QPoint(0, 0));
                        QPoint point = pointF.toPoint();
-                       point.rx() += item->width() / 2;
-                       point.ry() += item->height() / 2;
```
why was this removed?

brizental replied:
I left a lareg message on the commit, tl;dr; it's a workaround for tests to pass. See https://github.com/mozilla-mobile/mozilla-vpn-client/pull/8139/commits/6e546828ca46a3ada5860c1b3814834b6f80dc8b

Gela commented on src/telemetry/metrics.yaml:
```diff
@@ -100,6 +103,105 @@ impression:
+    bugs:
+      - https://mozilla-hub.atlassian.net/browse/VPN-5251
+    data_reviews:
+      - TODO
```
I'm not fully aware of how telemetry logistics work but is this going to be changed before the PR is merged?

brizental replied:
Yes, but it's awaiting on other discussions. https://github.com/mozilla-mobile/mozilla-vpn-client/pull/8139#issuecomment-1755563514

Gela commented on tests/functional/testInitialize.js:
```diff
@@ -78,6 +78,146 @@ describe('Initialize', function() {
+  });
+
+  describe('initialize related telemetry tests', () => {
+    if(vpn.runningOnWasm()) {
```
Oh this is neat and much more readable than what I've been doing!!
```js
if (!this.ctx.wasm) {
//...
}
```

Gela commented on tests/functional/testSubscriptionNeeded.js:
```diff
@@ -0,0 +1,172 @@
+ const { startAndConnect } = require('./setupVpn.js');
+
+describe("subscription needed tests", function() {
+  // TODO: Create tests for this screen.
```
Is there a follow up issue for this TODO? Or are you intending on adding additional tests later in this PR?

brizental replied:
Good call. I'll file a ticket.

Gela commented on tests/functional/testSubscriptionNeeded.js:
```diff
@@ -0,0 +1,172 @@
+              max_devices: 5,
+              // This is the important bit.
+              // The user data has no active subscriptions.
+              subscriptions: {},
```
For a user with no active subscription, should this be changed to the following instead?

```js
      subscriptions: {vpn: {active: false}},
```

brizental replied:
I think both result in the same effect? cc @jswinarton just in case. I think the difference between the two is that one is a user that has never subscribed and the other a user that was subscribed before and the subscription expired.

brizental commented on src/ui/authenticationInApp/ViewAuthenticationVerificationSessionByEmailNeeded.qml:
```diff
@@ -41,25 +56,44 @@ MZInAppAuthenticationBase {
+            objectName: _viewObjectName + "-cancel"
             anchors.horizontalCenter: parent.horizontalCenter
             onClicked: {
+                Glean.interaction.authenticationAborted.record({
```
Re-authentication is not really contemplated by the Miro boards. I left a question there though, I think it fell through the cracks. I'll ping Santiago and Yeonjoo again.

brizental replied:
I think for now it's safe to say we are ignoring re-auth though. We haven't recently done any scope changes to our FxA integration and IIUC that means re-auth won't happen often. This telemetry is about login and subscription and re-auth is a separate flow fom those two... So even if we decide to instrument re-auth flow that would be out of scope here.

brizental commented on src/authenticationinapp/authenticationinapp.cpp:
```diff
@@ -52,35 +52,6 @@ void AuthenticationInApp::setState(State state,
-    default:
-      logger.error()
-          << "Glean samples and Auth-in-app session types are out of sync";
-  }
```
These metrisc are now being recorded on the front-end.

brizental commented:
>  A PR of this length deserves a PR description that is more than copy/pasting from the ticket. It's kind to the reviewers to give a bit of background on the code, things to watch out for, some background on decisions that you made, etc.

I am not really sure what you are asking for. This a very mechanical pull request where I implement the metrics defined in the Miro board. I have taken care to make the git history here very semantic and there are explanations on the bodies of the commit messages that explain why some seemingly unrelated changes where necessary. I have added a note to the top of the description of this Pull Request that I recommend reviewing this commit by commit and using the git history as a reference. If there is something specific that is still unclear, please let me know.

brizental commented on tests/functional/testAuthenticationInApp.js:
```diff
@@ -1027,4 +1027,885 @@ describe('User authentication', function() {
+        // Leaving this for documentation more than anything.
+        // The reason why it's skipped is because we can't query the "Terms of Service"
+        // or the "Privacy Policy" links, because they are not actual QML components.
+        // They are HTML tags inside of a QML text component???
```
Calling on the QML experts to let me know if it is actually possible to query these elements cc @lesleyjanenorton @MattLichtenstein 

Essentially, what I'd need to here is query these `a` tags declared here https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/src/translations/strings.yaml#L689 with QML path.

brizental replied:
If I had the coordinates I could just use the click function though. The issue here is that I have no coordinates :/

MattLichtenstein replied:
Would it make sense to have the functional test click on hard-coded coordinates, and if the link is not opened, just skip the telemetry verification? If I understand correctly, this test is strictly testing telemetry on link click, so we can do our best to tell the test where the link is

brizental replied:
I guess in that case I'd prefer to just not test it and leave it as is for documentation purposes. It would be too fragile to add hard coded coordinates. WDYT?

brizental commented on src/ui/screens/ScreenSubscriptionInProgressWeb.qml:
```diff
@@ -10,6 +10,17 @@ MZLoader {
     objectName: "subscriptionInProgressWeb"
     headlineText: MZI18n.PurchaseWebInProgress2
 
+    // Hack: The Component.OnCompleted event was not working for me (Bea).
+    // So I am running this on property initialization.
```
@lesleyjanenorton @MattLichtenstein Do you know why the Component.onCompleted would not trigger here?

MattLichtenstein replied:
missing an `import QtQuick 2.15` at the top 🙂

MattLichtenstein commented on src/ui/screens/initialize/ViewInitialize.qml:
```diff
@@ -99,6 +99,16 @@ Item {
+                    action: "impression",
+                });
+
+                previousTelemetryScreenId = telemetryScreenId
```
where is `previousTelemetryScreenId` declared?

brizental replied:
Ha, that is deelopment leftover. Thanks for catching it, Matt.

mcleinman commented:
> > A PR of this length deserves a PR description that is more than copy/pasting from the ticket. It's kind to the reviewers to give a bit of background on the code, things to watch out for, some background on decisions that you made, etc.
> 
> I am not really sure what you are asking for. This a very mechanical pull request where I implement the metrics defined in the Miro board. I have taken care to make the git history here very semantic and there are explanations on the bodies of the commit messages that explain why some seemingly unrelated changes were necessary. I have added a note to the top of the description of this Pull Request that I recommend reviewing this commit by commit and using the git history as a reference. If there is something specific that is still unclear, please let me know.

This PR is a multithousand line change, and it includes a bunch of small things that don't (on their face) seem immediately connected to metrics (the change in `navigator.cpp`, the change in `inspectorhandler.cpp`, the movement in `ViewInitialize.qml`, the reference in a commit to a Jira ticket that seems unrelated to the commit, etc.). If you're not sure what I'm asking for, I'd ask that you look at other PRs put up in the repo - In a quick check of the last many PRs from our team, I can't find any PR with more than a couple lines that doesn't have some an explanation/description. Not everyone reviews PRs by looking at the commit history, and descriptions help reviewers get up to speed on a PR without having to dig deep into the preceding tickets/discussion. 

(While I don't feel that we all necessarily need to share this philosophy, I also think that if there are code changes that only make sense when you look at the commit messages AND they are not core to this PR's ticket, those commits are perhaps good candidates for being put up in a separate PR. It helps with better reviews, and leaving our main branch's commit history a bit cleaner - since these commit messages get buried once this PR gets merged.)

lesleyjanenorton commented:
@brizental @mcleinman 

> While I don't feel that we all necessarily need to share this philosophy

But a shared policy describing what information is expected when submitting PRs would help us here. I've added an agenda item to the Tuesday eng meeting so we can discuss with the broader team.

brizental commented:
> This PR is a multithousand line change, and it includes a bunch of small things that don't (on their face) seem immediately connected to metrics (the change in navigator.cpp, the change in inspectorhandler.cpp, the movement in ViewInitialize.qml, the reference in a commit to a Jira ticket that seems unrelated to the commit, etc.)
 
1. The change in `navigator.cpp` is isolated in it's own [commit](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/8139/commits/a4456636314c9a25adefc892fd7a0fe6adbc1121) with attached explanation. Here is a copy of said commit description:

> [VPN-5596](https://mozilla-hub.atlassian.net/browse/VPN-5596) - Skip loading of currently loaded screen
>
> Well, this was an issue during most of the time I was implementing
> the initialize screen events. Then right at the end I changed the
> approach and it was not an issue anymore. Still, here goes a freebie
> minor issues fix.

2. The changes to `inspectorhandler.cpp` are also isolated in their own [commit](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/8139/commits/b0bbac95f0e91ec01d915b67d4d2671bce4f256d) with attached explanation. Here is a copy of said commit description.

> Enable sign in cancel button click test
> 
> Ok, this is just a bit hacky. The test was failing because that button is
> below the fold. We'd have to scroll down to actually click on it. However,
> I cannot figure out how to scroll down for the life of me. I talked to Matt L.
> and he showed me the fun fact that if you click right on the fold without
> scrolling turns out you already reach the cancel button.
> 
> Now, tests are clicking in the middle of elements. So what I did is I changed
> the test to actually click at the top right corner of the element. In practice,
> this makes no difference. So instead of embarking in yet another rabbit hole
> to fix this, I refrained.

3. The changes to `ViewInitialize.qml` are not unrelated to the telemetry being recorded. Either way all of the initialize screen implementation is also isolated in it's own commit, which should aid reviewers in making the connections between these changes. See https://github.com/mozilla-mobile/mozilla-vpn-client/pull/8139/commits/6a27723dca1a58a4e1d31e0bba61a486705a1bf5#diff-ab7267df83e6eba12759e993c012fd6d9f4a45c389f87f2329c6bf4989cc4f19.

> If you're not sure what I'm asking for, I'd ask that you look at other PRs put up in the repo - In a quick check of the last many PRs from our team, I can't find any PR with more than a couple lines that doesn't have some an explanation/description.

Given the above, I feel like I have added review notes just not in the description exactly. I am still not sure what else is required here. What is still unclear? Do you want me to add the above information to the description? 

> Not everyone reviews PRs by looking at the commit history, and descriptions help reviewers get up to speed on a PR without having to dig deep into the preceding tickets/discussion.

I can see this not being a custom in the team. But the note I added to the description where I ask reviewers to review this commit by commit should address it, no?

>  I also think that if there are code changes that only make sense when you look at the commit messages AND they are not core to this PR's ticket, those commits are perhaps good candidates for being put up in a separate PR

Sure, I agree. Is this what you are asking for? I am happy to spin out the first commit, which is the unrelated one, into a seprate PR.

---

> But a shared policy describing what information is expected when submitting PRs would help us here. I've added an agenda item to the Tuesday eng meeting so we can discuss with the broader team.

Thanks @lesleyjanenorton. If possible, let's keep the policy discussions for that forum and focus on this specific Pull Request for now.

mcleinman reviewed (commented):
Leaving comment to clear the "requested changes".

Gela commented on nebula/ui/components/MZSignOut.qml:
```diff
@@ -9,7 +9,12 @@ MZFooterLink {
     onClicked: () => {
-                   VPNController.logout();
-               }
+        preLogoutCallback();
```
what purpose does `preLogoutCallback();` serve here, is it some sort of a delay?

brizental replied:
It's used to guarantee a certain block of code is run _before_ logout is called.

On CI tests were failing because the metric recording I implemented was not being called before logout destructed the view where the onClick handler was defined. See also: https://github.com/mozilla-mobile/mozilla-vpn-client/pull/8139/commits/769e0481e4212401913e126c3ea77e32c26b8700

brizental commented:
Merging this in despite of the MacOS build failures as they are unrelated. See: https://mozilla-hub.atlassian.net/browse/VPN-5818
